lambda_function.py

```diff
@@ -21,42 +21,49 @@ def lambda_handler(event, lambda_context):
     When multiple user IDs are entered,
-    each userID and its valid flag is inserted as a seperate row in BigQuery.
+    each row is created with a separate user ID's information and added to an array. This array, containing different rows, is added to the table.
     """
     messageBody = event["Records"]
 
     for eachMessage in messageBody:
         if "Sns" in eachMessage and "Message" in eachMessage["Sns"]:
-            # extracting the body of the message
             message = json.loads(eachMessage["Sns"]["Message"])[0]
 
-            # contains all the keys (or fields) and their values to be added into BigQuery.
-            row = {}
-
             # check if the key values exist in the message sent
             # even if one field is missing,
             #   the row isn't inserted and an error is logged
-
             if all(
-                (k in message for k in ("dateTime", "purpose", "authType", "user"))
+                (k in message for k in ("dateTime", "purpose", "authType",
+                 "user", "program", "userType", "sessionId", "userData"))
                 and (k in message["purpose"] for k in ("type", "subType", "params"))
                 and (k in message["user"] for k in ("values"))
                 and (k in message["purpose"]["params"] for k in ("platform", "id"))
+
             ):
-                # parsing through each user ID and its respective valid flag
-                for each in message["user"]["values"]:
-                    # the key values are the column names in the BQ table.
-                    row["timestamp"] = message["dateTime"]
+
+                user_values = message["user"]["values"]
+                user_values_length = len(user_values)
+                print(message)
+                for i in range(user_values_length):
+                    row = {}
+
+                    row["attendance_timestamp"] = message["dateTime"]
                     row["purpose_type"] = message["purpose"]["type"]
-                    row["purpose_sub_type"] = message["purpose"]["subType"]
+                    row["purpose_subtype"] = message["purpose"]["subType"]
                     row["platform"] = message["purpose"]["params"]["platform"]
                     row["platform_id"] = message["purpose"]["params"]["id"]
                     row["auth_type"] = message["authType"]
-                    row["user_id"] = each["userID"]
-                    row["user_data_validated"] = each["valid"]
-
-                    return insert_data(row)
+                    row["user_id"] = user_values[i]["userID"]
+                    row["user_data_validated"] = user_values[i]["valid"]
+                    row["number_multiple_entries"] = user_values_length
+                    row["program"] = message["group"]
+                    row["userType"] = message["userType"]
+                    row["session_id"] = message["sessionId"]
+                    row["user_ip_address"] = message["userNetworkData"]["userIp"]
```

Comment:
Ah, maybe just pass only the IP address here since that's all that's needed?

Reply:
yeah even I am a little confused, this PR is then not JUST for the IP address logging right?
All these other variables will also be logged?

Reply:
So there was no CD process in place before. I was deploying it manually from my computer. So this repo was a little outdated. I have created a PR for the CD process to keep the code up to date.

Reply:
Ohh got it